### PR TITLE
Re-register provider endpoints with admin interface after updates

### PR DESCRIPTION
**Pull request by chrisgresty**
See #555

### CODE REVIEW

File: components/proxy/src/main/java/com/hotels/styx/admin/handlers/ProviderRoutingHandler.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    private static UrlPatternRouter buildRouter(ObjectStore<ProviderObjectRecord> db) {
+        UrlPatternRouter.Builder routeBuilder = new UrlPatternRouter.Builder()
+                .get("/admin/" + ADMIN_ROOT, new ProviderListHandler(db));
+        db.entrySet().forEach(entry -> {
+                String providerName = entry.getKey();
+                entry.getValue().getStyxService().adminInterfaceHandlers(providerPath(providerName))
+                        .forEach((relPath, handler) ->
+                            routeBuilder.get(endpointPath(providerName, relPath), new HttpStreamer(MEGABYTE, handler))
```
Comment (mikkokar) [inline components/proxy/src/main/java/com/hotels/styx/admin/handlers/ProviderRoutingHandler.java]: We could simplify this code a bit by tweaking the `UrlpatternRouter`. So if `UrlPatternRouter` took a path prefix constructor argument, in this case `/admin/providers`, then all the subsequent calls to `routeBuilder.get` could simply pass in the relative path.

I think that would simplify this code a bit and we could possibly get rid of `providerPath` and `endpointPath` methods.

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/AdminInterfaceSpec.kt [hunk 2]
```diff
@@ -21,6 +21,7 @@ import io.kotlintest.Spec
 import io.kotlintest.matchers.string.shouldContain
 import io.kotlintest.matchers.string.shouldMatch
 import io.kotlintest.specs.FeatureSpec
+import java.io.File
```
Comment (mikkokar) [inline system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/AdminInterfaceSpec.kt]: Unused import ^^^.

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/AdminInterfaceSpec.kt [hunk 3]
```diff
@@ -64,6 +65,7 @@ class AdminInterfaceSpec : FeatureSpec() {
                         .shouldMatch("\"[0-9]{1,2}d [0-9]{1,2}h [0-9]{1,2}m\"".toRegex())
             }
         }
+
```
Comment (mikkokar) [inline system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/AdminInterfaceSpec.kt]: Unnecessary change.

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt [hunk 4]
```diff
@@ -90,8 +127,106 @@ class ProviderAdminInterfaceSpec : FeatureSpec() {
...
+                            .bodyAs(UTF_8)
+                    body shouldContain "/admin/providers/appA-monitor/status"
+                }
+            }
+
+            scenario("The new endpoint returns status information after server restart") {
+                Thread.sleep(5000)
+                val response = styxServer.adminRequest("/admin/providers/appA-monitor/status")
+                response.status() shouldBe OK
+                response.header(HttpHeaderNames.CONTENT_TYPE).get().toLowerCase() shouldBe HttpHeaderValues.APPLICATION_JSON.toString().toLowerCase()
```
Comment (mikkokar) [inline system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt]: Use static Imports for HTTP header names.

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt [hunk 5]
```diff
@@ -90,8 +127,106 @@ class ProviderAdminInterfaceSpec : FeatureSpec() {
...
+
+                eventually(5.seconds, AssertionError::class.java) {
+                    val body = styxServer.adminRequest("/admin/providers")
+                            .bodyAs(UTF_8)
+                    body shouldContain "/admin/providers/appA-monitor/status"
+                }
+            }
+
+            scenario("The new endpoint returns status information after server restart") {
+                Thread.sleep(5000)
```
Comment (mikkokar) [inline system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt]: Why the sleep here? Also 5 seconds is quite long. Is this justified?

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt [hunk 6]
```diff
@@ -90,8 +127,106 @@ class ProviderAdminInterfaceSpec : FeatureSpec() {
...
+                            .bodyAs(UTF_8)
+                    body shouldContain "/admin/providers/appA-monitor/status"
+                    body shouldContain "/admin/providers/appB-monitor/status"
+                }
+            }
+
+            scenario("The new endpoint returns status information without server restart") {
+                val response = styxServer.adminRequest("/admin/providers/appB-monitor/status")
+                response.status() shouldBe OK
+                response.header(HttpHeaderNames.CONTENT_TYPE).get().toLowerCase() shouldBe HttpHeaderValues.APPLICATION_JSON.toString().toLowerCase()
```
Comment (mikkokar) [inline system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt]: Use static imports.

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt (hunk 5 above)
Comment (chrisgresty) [reply system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt]: No, it's not needed at all. It was left over from investigating something earlier in development.

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt [hunk 7]
```diff
@@ -92,6 +120,59 @@ class ProviderAdminInterfaceSpec : FeatureSpec() {
                 body shouldContain "/admin/providers/mySecondMonitor/status"
             }
         }
+
+        feature("Endpoints for dynamically added Styx services are available in the Admin interface") {
+
+            scenario("Additional endpoints are listed on the Providers admin interface") {
+
+//                styxServer.components().servicesDatabase().insert("mockService",
+//                        ProviderObjectRecord("MockService", setOf(), mockk(), MockService("mockService")))
```
Comment (mikkokar) [inline system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt]: Some unnecessary comments here.

File: system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt [hunk 8]
```diff
@@ -92,6 +120,59 @@ class ProviderAdminInterfaceSpec : FeatureSpec() {
...
+                responseA.status() shouldBe NOT_FOUND
+                val responseB = styxServer.adminRequest("/admin/providers/appB-monitor/status")
+                responseB.status() shouldBe OK
+                responseB.header(CONTENT_TYPE).get().toLowerCase() shouldBe APPLICATION_JSON.toString().toLowerCase()
+                responseB.bodyAs(UTF_8) shouldBe "{ name: \"HealthCheckMonitoringService\" status: \"RUNNING\" }"
+            }
+
+            scenario("Endpoints for dynamically removed Styx services are not listed in the Admin interface") {
+
+//                styxServer.components().servicesDatabase().remove("mockService")
```
Comment (mikkokar) [inline system-tests/ft-suite/src/test/kotlin/com/hotels/styx/admin/ProviderAdminInterfaceSpec.kt]: ... and here.
